### PR TITLE
feat(sink): Add schema generator for argument sink types

### PR DESCRIPTION
**Pull request by JamesxX**
Implements schema generator for argument sink types, as described in #12.

To do:
- [x] Implement schema generator
- [x] Implement thorough testing
- [x] Add to documentation

### CODE REVIEW

File: tests/types/sink/test.typ [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+)
+#let sink-schema = z.sink(
+  positional: positional-schema,
+  named: named-schema
+)
+
+#let to-args-type(..args) = args
+= types/sink
+
+#z.parse(to-args-type("hello", named: "0"), sink-schema)
```
Comment (tingerrr) [inline tests/types/sink/test.typ:18]: Similar to my comments left on #18, we should likely discard the return value here.

**Review by tingerrr — CHANGES_REQUESTED** [review]
I think you forgot to format this, lots of inconsistent spacing.

File: tests/types/sink/test.typ (hunk 1 above)
Comment (JamesxX) [reply tests/types/sink/test.typ:16]: I'll hide these for the moment then

File: docs/manual.typ [hunk 2]
```diff
@@ -389,19 +389,36 @@ For the sake of brevity and owing to their consistency, the arguments that each
...
+]
+
 #command("choice", arg[choices], sarg[args], ret: "schema")[
   #argument("choices", types: "array")[Array of valid inputs]
 ]
 
 #pagebreak()
 
-#import("@preview/tidy:0.2.0")
+#import ("@preview/tidy:0.2.0")
```
Comment (tingerrr) [inline docs/manual.typ:415]:
```suggestion
#import "@preview/tidy:0.2.0"
```

File: src/types/sink.typ [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+#let to-args-type(..args) = args
+
+#let sink(
+  positional: none,
+  named: none,
+  ..args,
+) = {
+
+  if (positional != none) {
```
Comment (tingerrr) [inline src/types/sink.typ:13]:
```suggestion
  if positional != none {
```

File: src/types/sink.typ [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    assert-base-type(positional)
+  }
+  if (named != none) {
```
Comment (tingerrr) [inline src/types/sink.typ:16]:
```suggestion
  if named != none {
```

File: src/types/sink.typ [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    name: "argument-sink",
+    types: (arguments,),
+    ..args,
+  ) + (
+    positional-schema: positional,
+    named-schema: named,
+    handle-descendents: (self, it, ctx: z-ctx(), scope: ()) => {
+
+      let positional = it.pos()
+      if (self.positional-schema == none) {
```
Comment (tingerrr) [inline src/types/sink.typ:30]:
```suggestion
      if self.positional-schema == none {
```

File: src/types/sink.typ [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+        if (positional.len() > 0) {
```
Comment (tingerrr) [inline src/types/sink.typ:31]:
```suggestion
        if positional.len() > 0 {
```

File: src/types/sink.typ [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+        positional = (self.positional-schema.validate)(
+          self.positional-schema,
+          it.pos(),
+          ctx: ctx,
+          scope: (..scope, "positional"),
+        )
+      }
+
+      let named = it.named()
+      if (self.named-schema == none) {
```
Comment (tingerrr) [inline src/types/sink.typ:50]:
```suggestion
      if self.named-schema == none {
```

File: src/types/sink.typ [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+        if (named.len() > 0) {
```
Comment (tingerrr) [inline src/types/sink.typ:51]:
```suggestion
        if named.len() > 0 {
```

File: src/types/sink.typ [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+      } else {
+        named = (self.named-schema.validate)(
+          self.named-schema,
+          it.named(),
+          ctx: ctx,
+          scope: (..scope, "named"),
+        )
+      }
+
+      return to-args-type(..positional, ..named)
```
Comment (tingerrr) [inline src/types/sink.typ:69]:
```suggestion
      to-args-type(..positional, ..named)
```

**Review by tingerrr — COMMENTED** [review]
I've added a few suggestions that are not too problematic, I may have missed a few and I don't think they _need_ to be changed, but I find the personally unidiomatic.

I'm surprised the formatter doesn't impose those.

**Review by tingerrr — APPROVED** [review]
Looks good, can be merged, I think.

**Pull request merged**